Practical Example/IosInterviewPracticeProject/IosInterviewPracticeProject/Screens/CourseListVC.swift

```diff
@@ -0,0 +1,147 @@
+
+    switch sender.selectedSegmentIndex {
+    case 0:
+      filteredContent = courses
+    case 1:
+      filteredContent = courses.filter { $0.attributes.contentType == ContentType.article }
+    case 2:
+      filteredContent = courses.filter { $0.attributes.contentType == ContentType.collection }
+    default:
+      filteredContent = courses
+    }
+
+    DispatchQueue.main.async {
+      self.createSnapshot(from: filteredContent.sorted { $0.attributes.releasedAt > $1.attributes.releasedAt })
+    }
+  }
+
+    override func viewDidLoad() {
+      super.viewDidLoad()
+      courseList.delegate = self
+      courseList.dataSource = dataSource
+      configureDataSource()
+      fetchContent()
+    }
+
+  private func configureDataSource() {
+    dataSource = UITableViewDiffableDataSource<Section, Item>(tableView: courseList) { tableView, indexPath, item -> UITableViewCell in
+      guard let cell = tableView.dequeueReusableCell(withIdentifier: CourseListCell.reuseId, for: indexPath) as? CourseListCell else { return UITableViewCell() }
+      let duration = self.timeFormatter.string(from: TimeInterval(Double(item.attributes.duration))) ?? ""
+
+      cell.courseName.text = item.attributes.name
+      cell.courseDescription.text = item.attributes.descriptionPlainText
+      cell.duration.text = "(\(duration))"
+      cell.courseType.text = item.attributes.contentType == ContentType.article ? "Article Course" : "Video Course"
+      cell.releaseDate.text = item.attributes.releasedAt.formatDateString()//.description.formatDate()
+      cell.links = item.links
+
+      if let cardArtworkUrl = URL(string: item.attributes.cardArtworkUrl) {
+        CourseAPIService.shared.fetchCardArtwork(for: cardArtworkUrl) { result in
+          switch result {
+          case .success(let artwork):
+            DispatchQueue.main.async {
+              cell.artwork.contentMode = .scaleAspectFill
+              cell.artwork.layer.cornerRadius = 5
+              cell.artwork.layer.masksToBounds = true
+              cell.artwork.image = artwork
+            }
+          case .failure(let error):
+            switch error {
+            case .imageDataError:
+              print(error.localizedDescription)
+            default:
+              print("Artwork Error: \(error.localizedDescription)")
+            }
+          }
+        }
+      }
+      return cell
+    }
+  }
+
+  private func createSnapshot(from courses: [Item]) {
+    var snapshot = NSDiffableDataSourceSnapshot<Section, Item>()
+    snapshot.appendSections([.main])
+    snapshot.appendItems(courses, toSection: .main)
+    dataSource?.apply(snapshot, animatingDifferences: true, completion: nil)
+  }
+
+  private func fetchContent() {
+    let articleUrl = "https://raw.githubusercontent.com/raywenderlich/ios-interview/master/Practical%20Example/articles.json"
+    let videoUrl = "https://raw.githubusercontent.com/raywenderlich/ios-interview/master/Practical%20Example/videos.json"
```

Comment:
This should probably be defined in the API service as constants, and not in the view controller. You wouldn't want to have to modify your view controllers if the server URL changes.

Reply:
fixed